module/jobqueue/consumer.go

```diff
@@ -108,23 +110,28 @@ func (c *Consumer) Start(defaultIndex int64) error {
 }
 
 // Stop stops consuming jobs from the job queue.
+// It blocks until the existing worker finish processing the job
 // Note, it won't stop the existing worker from finishing their job
 func (c *Consumer) Stop() {
-	c.Lock()
-	defer c.Unlock()
-
+	c.mu.Lock()
 	c.running = false
+	// not to use `defer`, otherwise runningJobs.Wait will hold the lock and cause deadlock
+	c.mu.Unlock()
+
+	c.log.Info().Msg("stopping consumer")
+	c.runningJobs.Wait()
 	c.log.Info().Msg("consumer stopped")
 }
 
 // NotifyJobIsDone let the consumer know a job has been finished, so that consumer will take
 // the next job from the job queue if there are workers available
-func (c *Consumer) NotifyJobIsDone(jobID JobID) {
-	c.Lock()
-	defer c.Unlock()
+func (c *Consumer) NotifyJobIsDone(jobID module.JobID) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
 	c.log.Debug().Str("job_id", string(jobID)).Msg("finishing job")
 
 	if c.doneJob(jobID) {
+		c.runningJobs.Done()
```

Comment:
Do we would want to do this in `run` rather than here? We want to block `Stop` until all the goroutines we have spawned return. This will block `Stop` until all the goroutines we have spawned return _and_ all the pending jobs have been marked as finished, which could take quite a bit longer.

Reply:
I thought the graceful shutdown behavior you were thinking asis to wait until all pending jobs are finished. If you meant graceful is just for the call to `run` to finish, then we already have it, because calling `Stop` will attempts to acquire the same mutex lock as `run` does. If `run` hasn't been finished, `Stop` will be blocked and wait until `run` release the lock. And there will be at most 1 gorountine calling `run`, because of the lock.

Reply:
> If you meant graceful is just for the call to run to finish, then we already have it

Not for the call to `run` to finish, for goroutines _spawned_ in the call to `run` to finish. (i.e. those goroutines which are spawned [here](https://github.com/onflow/flow-go/pull/330/files#diff-2731207c7c63d225021f36118b9379bcb3e4394d9ef95e314020a55ff0715500R208))

Reply:
I see. Fixed here 
https://github.com/onflow/flow-go/pull/361/commits/3f4c9043a6f784b8f558fccbf8e00e7350b32eae